File: test/integration/testdata/sdk-smoke-python/smoke.py
"""
SDK acceptance smoke test for the stigmer Python SDK.

Exercises Agent CRUD + error handling (Tier 1) and optionally
workflow execution lifecycle (Tier 2) against a local Stigmer
service running in test mode.

Outputs a JSON result to stdout for the Go test orchestrator.
Diagnostic logs go to stderr.
"""

from __future__ import annotations

import json
import os
import sys
import time


def log(msg: str) -> None:
    print(f"[py-smoke] {msg}", file=sys.stderr, flush=True)


def main() -> None:
    result: dict[str, object] = {"tier1": "skip", "tier2": "skip", "errors": []}

    addr = os.environ.get("STIGMER_GRPC_ADDRESS", "")
    if not addr:
        result["errors"] = ["STIGMER_GRPC_ADDRESS not set"]  # type: ignore[assignment]
        print(json.dumps(result))
        sys.exit(1)

    workflow_runner_available = (
        os.environ.get("STIGMER_WORKFLOW_RUNNER_AVAILABLE", "false") == "true"
    )

    from stigmer import StigmerClient, is_not_found
    from stigmer import (
        AgentInput,
        WorkflowInput,
        WorkflowDocumentInput,
        WorkflowTaskInput,
        ExportInput,
        WorkflowExecutionInput,
    )
    from ai.stigmer.agentic.workflow.v1 import enum_pb2 as wf_enum
    from ai.stigmer.agentic.workflowexecution.v1 import enum_pb2 as wfe_enum

    client = StigmerClient(api_key="test-api-key", base_url=addr, insecure=True)

    # -------------------------------------------------------------------
    # Tier 1: Agent CRUD + error handling
    # -------------------------------------------------------------------
    try:
        agent_name = f"sdk-smoke-py-{int(time.time() * 1000)}"

        # Create
        log("creating agent...")
        created = client.agents.apply(
            AgentInput(
                name=agent_name,
                org="test-org",
                description="SDK acceptance smoke test agent (Python)",
                instructions="You are a test agent. Respond with exactly: hello from sdk smoke test",
            )
        )
        agent_id = created.metadata.id
        if not agent_id:
            raise ValueError("created agent has no ID")
        log(f"created agent: id={agent_id}")

        # Get
        log("fetching agent...")
        fetched = client.agents.get(agent_id)
        assert_equal("agent name", fetched.metadata.name, agent_name)
        assert_equal("agent org", fetched.metadata.org, "test-org")
        assert_equal(
            "agent description",
            fetched.spec.description,
            "SDK acceptance smoke test agent (Python)",
        )

        # List
        log("listing agents...")
        from stigmer import ListParams

        list_result = client.agents.list(ListParams(org="test-org"))
        if list_result.total_count < 1:
            raise ValueError(
                f"agent list must contain at least one entry, got {list_result.total_count}"
            )

        # Delete
        log("deleting agent...")
        client.agents.delete(agent_id)

        # Get deleted -> NOT_FOUND
        log("verifying NOT_FOUND after delete...")
        try:
            client.agents.get(agent_id)
            raise ValueError("expected NOT_FOUND error but get succeeded")
        except Exception as err:
            if not is_not_found(err):
                raise ValueError(f"expected NOT_FOUND error, got: {err}") from err

        log("Tier 1 passed")
        result["tier1"] = "pass"

    except Exception as err:
        log(f"Tier 1 FAILED: {err}")
        result["tier1"] = "fail"
        result["errors"].append(f"tier1: {err}")  # type: ignore[union-attr]
        print(json.dumps(result))
        sys.exit(1)

    # -------------------------------------------------------------------
    # Tier 2: Workflow execution lifecycle
    # -------------------------------------------------------------------
    if not workflow_runner_available:
        log("workflow-runner not available — skipping Tier 2")
        result["tier2"] = "skip"
        print(json.dumps(result))
        sys.exit(0)

    try:
        workflow_name = f"sdk-smoke-wf-py-{int(time.time() * 1000)}"

        # Apply workflow
        log("applying workflow...")
        applied = client.workflows.apply(
            WorkflowInput(
                name=workflow_name,
                org="test-org",
                document=WorkflowDocumentInput(
                    dsl="1.0.0",
                    namespace="test-org",
                    name=workflow_name,
                    version="1.0.0",
                ),
                tasks=[
                    WorkflowTaskInput(
                        name="setGreeting",
                        kind=wf_enum.WorkflowTaskKind.set_vars,
                        task_config={
                            "variables": {"greeting": "hello-from-py-sdk-smoke-test"}
                        },
                        export=ExportInput(as_="${.}"),
                    ),
                ],
            )
        )
        workflow_id = applied.metadata.id
        if not workflow_id:
            raise ValueError("applied workflow has no ID")
        log(f"applied workflow: id={workflow_id}")

        # Create execution
        log("creating workflow execution...")
        execution = client.workflow_executions.create(
            WorkflowExecutionInput(
                name=f"sdk-smoke-exec-py-{int(time.time() * 1000)}",
                org="test-org",
                workflow_id=workflow_id,
                trigger_message="SDK acceptance smoke test",
            )
        )
        execution_id = execution.metadata.id
        if not execution_id:
            raise ValueError("created execution has no ID")
        log(f"created execution: id={execution_id}")

        # Poll until COMPLETED (90s timeout, 2s interval)
        deadline = time.time() + 90
        last_phase = ""
        completed = False

        while time.time() < deadline:
            time.sleep(2)
            fetched_exec = client.workflow_executions.get(execution_id)
            phase = fetched_exec.status.phase
            last_phase = wfe_enum.ExecutionPhase.Name(phase)

            if phase == wfe_enum.ExecutionPhase.EXECUTION_COMPLETED:
                log(f"execution completed: id={execution_id}")

                for task in fetched_exec.status.tasks:
                    if task.task_name == "setGreeting":
                        assert_equal(
                            "task status",
                            task.status,
                            wfe_enum.WorkflowTaskStatus.WORKFLOW_TASK_COMPLETED,
                        )
                completed = True
                break

            if phase in (
                wfe_enum.ExecutionPhase.EXECUTION_FAILED,
                wfe_enum.ExecutionPhase.EXECUTION_CANCELLED,
            ):
                raise ValueError(
                    f"execution reached terminal failure phase: {last_phase}"
                )

        if not completed:
            raise TimeoutError(
                f"timed out waiting for execution to complete; last phase: {last_phase}"
            )

        # Cleanup
        client.workflows.delete(workflow_id)

        log("Tier 2 passed")
        result["tier2"] = "pass"

    except Exception as err:
        log(f"Tier 2 FAILED: {err}")
        result["tier2"] = "fail"
        result["errors"].append(f"tier2: {err}")  # type: ignore[union-attr]
        print(json.dumps(result))
        sys.exit(1)

    print(json.dumps(result))
    sys.exit(0)


def assert_equal(label: str, actual: object, expected: object) -> None:
    if actual != expected:
        raise ValueError(f"{label} mismatch: expected {expected!r}, got {actual!r}")


if __name__ == "__main__":
    main()

File: test/integration/testdata/sdk-smoke-ts/smoke.ts
/**
 * SDK acceptance smoke test for @stigmer/sdk (TypeScript).
 *
 * Exercises Agent CRUD + error handling (Tier 1) and optionally
 * workflow execution lifecycle (Tier 2) against a local Stigmer
 * service running in test mode.
 *
 * Outputs a JSON result to stdout for the Go test orchestrator.
 * Diagnostic logs go to stderr.
 */

import { createGrpcTransport } from "@connectrpc/connect-node";
import { Stigmer, isNotFound } from "@stigmer/sdk";
import { WorkflowTaskKind } from "@stigmer/protos/ai/stigmer/agentic/workflow/v1/enum_pb";
import { ExecutionPhase, WorkflowTaskStatus } from "@stigmer/protos/ai/stigmer/agentic/workflowexecution/v1/enum_pb";

interface SmokeResult {
  tier1: "pass" | "fail" | "skip";
  tier2: "pass" | "fail" | "skip";
  errors: string[];
}

const result: SmokeResult = { tier1: "skip", tier2: "skip", errors: [] };

function log(msg: string): void {
  process.stderr.write(`[ts-smoke] ${msg}\n`);
}

function finish(exitCode: number): never {
  process.stdout.write(JSON.stringify(result) + "\n");
  process.exit(exitCode);
}

async function main(): Promise<void> {
  const addr = process.env.STIGMER_GRPC_ADDRESS;
  if (!addr) {
    result.errors.push("STIGMER_GRPC_ADDRESS not set");
    finish(1);
  }

  const workflowRunnerAvailable =
    process.env.STIGMER_WORKFLOW_RUNNER_AVAILABLE === "true";

  // connect-node v2's gRPC transport is HTTP/2-only; the v1-era httpVersion
  // option no longer exists.
  const transport = createGrpcTransport({
    baseUrl: `http://${addr}`,
  });

  const stigmer = new Stigmer({
    baseUrl: `http://${addr}`,
    customTransport: transport,
  });

  // -----------------------------------------------------------------------
  // Tier 1: Agent CRUD + error handling
  // -----------------------------------------------------------------------
  try {
    const agentName = `sdk-smoke-ts-${Date.now()}`;

    // Create
    log("creating agent...");
    const created = await stigmer.agent.apply({
      name: agentName,
      org: "test-org",
      description: "SDK acceptance smoke test agent (TypeScript)",
      instructions:
        "You are a test agent. Respond with exactly: hello from sdk smoke test",
    });

    const agentId = created.metadata?.id;
    if (!agentId) throw new Error("created agent has no ID");
    log(`created agent: id=${agentId}`);

    // Get
    log("fetching agent...");
    const fetched = await stigmer.agent.get(agentId);
    assertEqual("agent name", fetched.metadata?.name, agentName);
    assertEqual("agent org", fetched.metadata?.org, "test-org");
    assertEqual(
      "agent description",
      fetched.spec?.description,
      "SDK acceptance smoke test agent (TypeScript)",
    );

    // List
    log("listing agents...");
    const listResult = await stigmer.agent.list({ org: "test-org" });
    if (listResult.totalCount < 1) {
      throw new Error(
        `agent list must contain at least one entry, got ${listResult.totalCount}`,
      );
    }

    // Delete
    log("deleting agent...");
    await stigmer.agent.delete(agentId);

    // Get deleted -> NOT_FOUND
    log("verifying NOT_FOUND after delete...");
    try {
      await stigmer.agent.get(agentId);
      throw new Error("expected NOT_FOUND error but get succeeded");
    } catch (err) {
      if (!isNotFound(err)) {
        throw new Error(`expected NOT_FOUND error, got: ${err}`);
      }
    }

    log("Tier 1 passed");
    result.tier1 = "pass";
  } catch (err) {
    const msg = err instanceof Error ? err.message : String(err);
    log(`Tier 1 FAILED: ${msg}`);
    result.tier1 = "fail";
    result.errors.push(`tier1: ${msg}`);
    finish(1);
  }

  // -----------------------------------------------------------------------
  // Tier 2: Workflow execution lifecycle
  // -----------------------------------------------------------------------
  if (!workflowRunnerAvailable) {
    log("unified runner not available — skipping Tier 2");
    result.tier2 = "skip";
    finish(0);
  }

  try {
    const workflowName = `sdk-smoke-wf-ts-${Date.now()}`;

    // Apply workflow
    log("applying workflow...");
    const applied = await stigmer.workflow.apply({
      name: workflowName,
      org: "test-org",
      document: {
        dsl: "1.0.0",
        namespace: "test-org",
        name: workflowName,
        version: "1.0.0",
      },
      tasks: [
        {
          name: "setGreeting",
          kind: WorkflowTaskKind.set_vars,
          taskConfig: {
            variables: { greeting: "hello-from-ts-sdk-smoke-test" },
          },
        },
      ],
    });

    const workflowId = applied.metadata?.id;
    if (!workflowId) throw new Error("applied workflow has no ID");
    log(`applied workflow: id=${workflowId}`);

    // Create execution
    log("creating workflow execution...");
    const execution = await stigmer.workflowExecution.create({
      name: `sdk-smoke-exec-ts-${Date.now()}`,
      org: "test-org",
      workflowId,
      triggerMessage: "SDK acceptance smoke test",
    });

    const executionId = execution.metadata?.id;
    if (!executionId) throw new Error("created execution has no ID");
    log(`created execution: id=${executionId}`);

    // Poll until COMPLETED (30s timeout, 2s interval)
    const deadline = Date.now() + 90_000;
    let lastPhase = "";
    while (Date.now() < deadline) {
      await sleep(2000);
      const fetched = await stigmer.workflowExecution.get(executionId);
      const phase = fetched.status?.phase;

      if (phase !== undefined) {
        lastPhase = ExecutionPhase[phase] ?? String(phase);
      }

      if (phase === ExecutionPhase.EXECUTION_COMPLETED) {
        log(`execution completed: id=${executionId}`);

        // Verify task status
        const tasks = fetched.status?.tasks ?? [];
        const setGreeting = tasks.find((t) => t.taskName === "setGreeting");
        if (setGreeting) {
          assertEqual(
            "task status",
            setGreeting.status,
            WorkflowTaskStatus.WORKFLOW_TASK_COMPLETED,
          );
        }

        break;
      }

      if (
        phase === ExecutionPhase.EXECUTION_FAILED ||
        phase === ExecutionPhase.EXECUTION_CANCELLED
      ) {
        throw new Error(`execution reached terminal failure phase: ${lastPhase}`);
      }
    }

    if (Date.now() >= deadline) {
      throw new Error(
        `timed out waiting for execution to complete; last phase: ${lastPhase}`,
      );
    }

    // Cleanup
    await stigmer.workflow.delete(workflowId);

    log("Tier 2 passed");
    result.tier2 = "pass";
  } catch (err) {
    const msg = err instanceof Error ? err.message : String(err);
    log(`Tier 2 FAILED: ${msg}`);
    result.tier2 = "fail";
    result.errors.push(`tier2: ${msg}`);
    finish(1);
  }

  finish(0);
}

function assertEqual<T>(label: string, actual: T, expected: T): void {
  if (actual !== expected) {
    throw new Error(
      `${label} mismatch: expected ${JSON.stringify(expected)}, got ${JSON.stringify(actual)}`,
    );
  }
}

function sleep(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

main().catch((err) => {
  result.errors.push(`unhandled: ${err instanceof Error ? err.message : String(err)}`);
  finish(1);
});
